README.md
# feedyard/circleci-base-agent [![CircleCI](https://circleci.com/gh/feedyard/circleci-base-agent.svg?style=svg)](https://circleci.com/gh/feedyard/circleci-base-agent) [![Docker Repository on Quay](https://quay.io/repository/feedyard/circleci-base-agent/status "Docker Repository on Quay")](https://quay.io/repository/feedyard/circleci-base-agent) [![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/feedyard/circleci-base-agent/master/LICENSE) [![Base Image](https://img.shields.io/badge/FROM-alpine-blue.svg)](https://alpinelinux.org)

Based on [feedyard/circleci-remote-docker](https://github.com/feedyard/circleci-remote-docker). Includes common tools for building and testing
docker images to be used as primary containers in infrastructure as code pipelines pipelines.

_additions_

apk/bin  | pip      |  gems
---------|----------|---------
bash     | invoke   | inspec
docker   | pylint   | rubocop
python3  | yamllint |
ruby     |          |
curl     |          |
wget     |          |
gnupg    |          |
openssl  |          |
jq       |          |
hadolint |          |

See CHANGELOG for list of installed packages/versions.


CHANGELOG.md
### 09-12-2019 release version 4.12.0

hadolint 1.17.2
inspec 4.16.0
docker 19.03.2-r0

### 08-28-2019 release version 4.11.0

FROM quay.io/feedyard/circleci-remote-docker:5.3.1

pip==19.2.3
setuptools==41.2.0
INSPEC_VERSION=4.12.0
INVOKE_VERSION=1.3.0
PYLINT_VERSION=2.3.1
YAMLLINT_VERSION=1.17.0
RUBOCOP_VERSION=0.74.0
HADOLINT_VERSION=1.17.1

http://dl-cdn.alpinelinux.org/alpine/v3.8/main
jq=1.6_rc1-r1

http://dl-cdn.alpinelinux.org/alpine/v3.10/main
bash=5.0.0-r0
openrc=0.41.2-r1
curl=7.65.1-r0
wget=1.20.3-r0
ruby=2.5.5-r0
ruby-bundler=2.0.2-r0
ruby-bigdecimal=2.5.5-r0
ruby-webrick=2.5.5-r0
gnupg=2.2.16-r0
openssl=1.1.1c-r0

http://dl-cdn.alpinelinux.org/alpine/edge/main
python3=3.7.4-r0

http://dl-cdn.alpinelinux.org/alpine/edge/community
docker=19.03.1-r2


### 06-19-2019 add and update packages

Installing openssl (1.1.1b-r1)
yamllint==1.16.0
installed inspec-4.6.4


### 06-07-2019 base image and package updates

feedyard/circleci-remote-docker:5.1.0
inspec-bin 4.3.2


### 06-03-2019 security patched package for jq

Installing jq (1.6_rc1-r1)


### 05-18-2019 package updates

wget=1.20.3-r0
ruby 2.5.5-r0
python3=3.6.8-r2

in removed build packages
ruby-dev=2.5.5-r0
gcc 2.3.0-r0
g++ 2.3.0-r0
python3-dev=3.6.8-r2

### 04-05-2019 publish version 4.5.0

(1/50) Installing readline (7.0.003-r1)
(2/50) Installing bash (4.4.19-r1)
(3/50) Installing curl (7.64.0-r1)
(4/50) Installing libseccomp (2.3.3-r1)
(5/50) Installing runc (1.0.0_rc6-r1)
(6/50) Installing containerd (1.2.2-r0)
(7/50) Installing libmnl (1.0.4-r0)
(8/50) Installing jansson (2.11-r0)
(9/50) Installing libnftnl-libs (1.1.1-r0)
(10/50) Installing iptables (1.6.2-r1)
(11/50) Installing tini-static (0.18.0-r0)
(12/50) Installing device-mapper-libs (2.02.182-r0)
(13/50) Installing libltdl (2.4.6-r5)
(14/50) Installing docker (18.09.1-r0)
(15/50) Installing libgpg-error (1.33-r0)
(16/50) Installing libassuan (2.5.1-r0)
(17/50) Installing libcap (2.26-r0)
(18/50) Installing pinentry (1.1.0-r0)
(19/50) Installing libbz2 (1.0.6-r6)
(20/50) Installing libgcrypt (1.8.4-r0)
(21/50) Installing gmp (6.1.2-r1)
(22/50) Installing nettle (3.4.1-r0)
(23/50) Installing libffi (3.2.1-r6)
(24/50) Installing p11-kit (0.23.14-r0)
(25/50) Installing libtasn1 (4.13-r0)
(26/50) Installing libunistring (0.9.10-r0)
(27/50) Installing gnutls (3.6.4-r0)
(28/50) Installing libksba (1.3.5-r0)
(29/50) Installing db (5.3.28-r1)
(30/50) Installing libsasl (2.1.27-r1)
(31/50) Installing libldap (2.4.47-r2)
(32/50) Installing npth (1.6-r0)
(33/50) Installing sqlite-libs (3.26.0-r3)
(34/50) Installing gnupg (2.2.12-r0)
(35/50) Installing oniguruma (6.9.1-r0)
(36/50) Installing jq (1.6-r0)
(37/50) Installing openrc (0.39.2-r3)
(38/50) Installing docker-openrc (18.09.1-r0)
(39/50) Installing gdbm (1.13-r1)
(40/50) Installing xz-libs (5.2.4-r0)
(41/50) Installing python3 (3.6.8-r1)
(42/50) Installing yaml (0.2.1-r0)
(43/50) Installing ruby-libs (2.5.3-r1)
(44/50) Installing ruby (2.5.3-r1)
(45/50) Installing ruby-bigdecimal (2.5.3-r1)
(46/50) Installing ruby-etc (2.5.3-r1)
(47/50) Installing ruby-io-console (2.5.3-r1)
(48/50) Installing ruby-bundler (1.17.1-r0)
(49/50) Installing ruby-webrick (2.5.3-r1)
(50/50) Installing wget (1.20.1-r0)

(1/25) Installing binutils (2.31.1-r2)
(2/25) Installing libmagic (5.35-r0)
(3/25) Installing file (5.35-r0)
(4/25) Installing isl (0.18-r0)
(5/25) Installing libgomp (8.2.0-r2)
(6/25) Installing libatomic (8.2.0-r2)
(7/25) Installing libgcc (8.2.0-r2)
(8/25) Installing mpfr3 (3.1.5-r1)
(9/25) Installing mpc1 (1.0.3-r1)
(10/25) Installing libstdc++ (8.2.0-r2)
(11/25) Installing gcc (8.2.0-r2)
(12/25) Installing musl-dev (1.1.20-r4)
(13/25) Installing libc-dev (0.7.1-r0)
(14/25) Installing g++ (8.2.0-r2)
(15/25) Installing make (4.2.1-r2)
(16/25) Installing fortify-headers (1.0-r0)
(17/25) Installing build-base (0.5-r1)
(18/25) Installing pkgconf (1.6.0-r0)
(19/25) Installing python3-dev (3.6.8-r1)
(20/25) Installing libgmpxx (6.1.2-r1)
(21/25) Installing gmp-dev (6.1.2-r1)
(22/25) Installing ruby-dev (2.5.3-r1)
(23/25) Installing linux-headers (4.18.13-r1)
(24/25) Installing libffi-dev (3.2.1-r6)
(25/25) Installing build-dependencies (0)

Successfully installed
pip-19.0.3
setuptools-41.0.0
astroid-2.2.5
invoke-1.2.0
isort-4.3.16
lazy-object-proxy-1.3.1
mccabe-0.6.1
pathspec-0.5.9
pylint-2.3.1
pyyaml-5.1
six-1.12.0
typed-ast-1.3.1
wrapt-1.11.1
yamllint-1.15.0

Successfully installed json-2.2.0
Successfully installed mixlib-shellout-2.4.4
Successfully installed net-ssh-5.2.0
Successfully installed net-scp-1.2.1
Successfully installed builder-3.2.3
Successfully installed erubis-2.7.0
Successfully installed ffi-1.10.0
Successfully installed gssapi-1.2.0
Successfully installed gyoku-1.3.1
Successfully installed httpclient-2.8.3
Successfully installed multi_json-1.13.1
Successfully installed little-plugger-1.1.4
Successfully installed logging-2.2.2
Successfully installed nori-2.6.0
Successfully installed rubyntlm-0.6.2
Successfully installed winrm-2.3.1
Successfully installed rubyzip-1.2.2
Successfully installed winrm-fs-1.3.2
Successfully installed excon-0.62.0
Successfully installed docker-api-1.34.2
Successfully installed jmespath-1.4.0
Successfully installed aws-eventstream-1.0.2
Successfully installed aws-sigv4-1.1.0
Successfully installed aws-sdk-core-2.11.256
Successfully installed aws-sdk-resources-2.11.256
Successfully installed aws-sdk-2.11.256
Successfully installed concurrent-ruby-1.1.5
Successfully installed multipart-post-2.0.0
Successfully installed faraday-0.15.4
Successfully installed unf_ext-0.0.7.5
Successfully installed unf-0.1.4
Successfully installed domain_name-0.5.20180417
Successfully installed http-cookie-1.0.3
Successfully installed faraday-cookie_jar-0.0.6
Successfully installed timeliness-0.4.0
Successfully installed ms_rest-0.7.3
Successfully installed ms_rest_azure-0.11.0
Successfully installed azure_mgmt_resources-0.17.3
Successfully installed azure_graph_rbac-0.17.1

profiles/circleci-base-agent/controls/circleci-base-image.rb
control 'packages' do
  impact 1.0
  title 'confirm package installation'
  desc 'confirm all desired packages are installed'
  describe command('apk info') do
    its('stdout') { should include ('curl') }
    its('stdout') { should include ('wget') }
    its('stdout') { should include ('python3') }
    its('stdout') { should include ('ruby') }
    its('stdout') { should include ('ruby-bundler') }
    its('stdout') { should include ('ruby-dev') }
    its('stdout') { should include ('g++') }
    its('stdout') { should include ('libffi-dev') }
    its('stdout') { should include ('musl-dev') }
    its('stdout') { should include ('make') }
    its('stdout') { should include ('jq') }
  end
end

control 'curl version' do
  impact 1.0
  title 'confirm curl version installed'
  desc 'confirm version reported by curl matches the desired version'
  describe command('curl -V') do
    its('stdout') { should include ('7.6') }
  end
end

control 'wget version' do
  impact 1.0
  title 'confirm wget version installed'
  desc 'confirm version reported by wget matches the desired version'
  describe command('wget --version') do
    its('stdout') { should include ('1.19') }
  end
end

control 'python3 version' do
  impact 1.0
  title 'confirm python3 version installed'
  desc 'confirm version reported by python3 matches the desired version'
  describe command('python3 -V') do
    its('stdout') { should include ('3.6') }
  end
end

control 'ruby version' do
  impact 1.0
  title 'confirm ruby version installed'
  desc 'confirm version reported by ruby matches the desired version'
  describe command('ruby -v') do
    its('stdout') { should include ('2.5') }
  end
end

control 'bundler version' do
  impact 1.0
  title 'confirm bundler version installed'
  desc 'confirm version reported by bundler matches the desired version'
  describe command('bundler -v') do
    its('stdout') { should include ('1.16') }
  end
end

control 'g++ version' do
  impact 1.0
  title 'confirm g++ version installed'
  desc 'confirm version reported by g++ matches the desired version'
  describe command('g++ --version') do
    its('stdout') { should include ('6.4') }
  end
end

control 'make version' do
  impact 1.0
  title 'confirm make version installed'
  desc 'confirm version reported by make matches the desired version'
  describe command('make -v') do
    its('stdout') { should include ('4.2.1') }
  end
end

control 'jq version' do
  impact 1.0
  title 'confirm jq version installed'
  desc 'confirm version reported by jq matches the desired version'
  describe command('jq --version') do
    its('stdout') { should include ('v3.7') }
  end
end

control 'pip version' do
  impact 1.0
  title 'confirm pip version installed'
  desc 'confirm version reported by pip3 matches the desired version'
  describe command('pip -V') do
    its('stdout') { should include ('18.0') }
  end
end

control 'python packages' do
  impact 1.0
  title 'confirm python package installation'
  desc 'confirm all desired python packages are installed'
  describe command('pip list') do
    its('stdout') { should include ('invoke') }
  end
end

control 'invoke version' do
  impact 1.0
  title 'confirm invoke version installed'
  desc 'confirm version reported by invoke matches the desired version'
  describe command('invoke -V') do
    its('stdout') { should include ('1.2') }
  end
end

control 'inspec version' do
  impact 1.0
  title 'confirm inspec version installed'
  desc 'confirm version reported by inspec matches the desired version'
  describe command('inspec version') do
    its('stdout') { should include ('2.2') }
  end
end


profiles/circleci-base-agent/controls/circleci-base-agent.rb
control 'packages' do
  impact 1.0
  title 'confirm package installation'
  desc 'confirm all desired packages are installed'
  describe command('apk info') do
    its('stdout') { should include ('bash') }
    its('stdout') { should include ('docker') }
    its('stdout') { should include ('openrc') }
    its('stdout') { should include ('curl') }
    its('stdout') { should include ('wget') }
    its('stdout') { should include ('jq') }
    its('stdout') { should include ('python3') }
    its('stdout') { should include ('ruby') }
    its('stdout') { should include ('ruby-bundler') }
    its('stdout') { should include ('ruby-webrick') }
    its('stdout') { should include ('ruby-bigdecimal') }
    its('stdout') { should include ('jq') }
    its('stdout') { should include ('gnupg') }
    its('stdout') { should include ('openssl') }
    its('stdout') { should_not include ('ruby-dev') }
    its('stdout') { should_not include ('g++') }
    its('stdout') { should_not include ('libffi-dev') }
    its('stdout') { should_not include ('musl-dev') }
    its('stdout') { should_not include ('make') }
  end
end

control 'docker version' do
  impact 1.0
  title 'confirm docker version installed'
  desc 'confirm version reported by docker matches the desired version'
  describe command('docker version') do
    its('stdout') { should include ('19.03') }
  end
end

control 'bash version' do
  impact 1.0
  title 'confirm bash version installed'
  desc 'confirm version reported by bash matches the desired version'
  describe command('bash --version') do
    its('stdout') { should include ('5.0') }
  end
end

control 'curl version' do
  impact 1.0
  title 'confirm curl version installed'
  desc 'confirm version reported by curl matches the desired version'
  describe command('curl -V') do
    its('stdout') { should include ('7.6') }
  end
end

control 'wget version' do
  impact 1.0
  title 'confirm wget version installed'
  desc 'confirm version reported by wget matches the desired version'
  describe command('wget --version') do
    its('stdout') { should include ('1.20') }
  end
end

control 'python3 version' do
  impact 1.0
  title 'confirm python3 version installed'
  desc 'confirm version reported by python3 matches the desired version'
  describe command('python3 -V') do
    its('stdout') { should include ('3.7') }
  end
end

control 'ruby version' do
  impact 1.0
  title 'confirm ruby version installed'
  desc 'confirm version reported by ruby matches the desired version'
  describe command('ruby -v') do
    its('stdout') { should include ('2.5') }
  end
end

control 'bundler version' do
  impact 1.0
  title 'confirm bundler version installed'
  desc 'confirm version reported by bundler matches the desired version'
  describe command('bundler -v') do
    its('stdout') { should include ('2.0') }
  end
end

control 'jq version' do
  impact 1.0
  title 'confirm jq version installed'
  desc 'confirm version reported by jq matches the desired version'
  describe command('jq --version') do
    its('stdout') { should include ('v3.7') }
  end
end

control 'pip version' do
  impact 1.0
  title 'confirm pip version installed'
  desc 'confirm version reported by pip3 matches the desired version'
  describe command('pip -V') do
    its('stdout') { should include ('19') }
  end
end

control 'python packages' do
  impact 1.0
  title 'confirm python package installation'
  desc 'confirm all desired python packages are installed'
  describe command('pip list') do
    its('stdout') { should include ('invoke') }
    its('stdout') { should include ('pylint') }
    its('stdout') { should include ('yamllint') }
  end
end

control 'invoke version' do
  impact 1.0
  title 'confirm invoke version installed'
  desc 'confirm version reported by invoke matches the desired version'
  describe command('invoke -V') do
    its('stdout') { should include ('1.3') }
  end
end

control 'pylint version' do
  impact 1.0
  title 'confirm pylint version installed'
  desc 'confirm version reported by pylint matches the desired version'
  describe command('pylint --version') do
    its('stdout') { should include ('2.3') }
  end
end

control 'yamllint version' do
  impact 1.0
  title 'confirm yamllint version installed'
  desc 'confirm version reported by yamllint matches the desired version'
  describe command('yamllint -v') do
    its('stdout') { should include ('1.17') }
  end
end

control 'rubocop version' do
  impact 1.0
  title 'confirm rubocop version installed'
  desc 'confirm version reported by rubocop matches the desired version'
  describe command('rubocop -v') do
    its('stdout') { should include ('0.74') }
  end
end

control 'inspec version' do
  impact 1.0
  title 'confirm inspec version installed'
  desc 'confirm version reported by inspec matches the desired version'
  describe command('inspec version') do
    its('stdout') { should include ('4.16') }
  end
end

control 'gnugp version' do
  impact 1.0
  title 'confirm gnugp version installed'
  desc 'confirm version reported by gnugp matches the desired version'
  describe command('gpg --version') do
    its('stdout') { should include ('2.2') }
  end
end

control 'openssl version' do
  impact 1.0
  title 'confirm openssl version installed'
  desc 'confirm version reported by gnugp matches the desired version'
  describe command('openssl version') do
    its('stdout') { should include ('1.1') }
  end
end
